#Library

In [ ]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.impute import SimpleImputer
!pip install pyjanitor
import janitor
import missingno as msno
import re

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
from sklearn.impute import KNNImputer
from sklearn.compose import ColumnTransformer, make_column_selector

In [ ]:
import warnings
warnings.filterwarnings('ignore', category=DeprecationWarning)

#DataFrame Read

In [5]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [ ]:
%cd /content/gdrive/MyDrive/DatasetFrame/Club Member Info/

In [8]:
df = pd.read_csv('club_member_info.csv', encoding='latin')
df.sample(20)

full_name   age martial_status                        email  \
861          sadie spybey  54.0        married     sspybeynu@technorati.com   
698       ardelis larmett  50.0         single     alarmettjc@blinklist.com   
260        Vinnie Eckford  41.0      separated  veckford77@surveymonkey.com   
1816       ???Kania Ginni  25.0        married           kginnimg@baidu.com   
1371       Worthy Proctor  50.0         single         wproctora6@amazon.de   
591         Tripp Mardell  46.0        married        tmardellgd@netlog.com   
835        BASIA DEARLOVE  36.0        married      bdearloven4@shop-pro.jp   
1744  Leonelle Beechcraft  41.0        married    lbeechcraftkg@oaic.gov.au   
1035         Lexi Gaffney  38.0         single      lgaffneyv@goodreads.com   
217       Alyson Dethloff  50.0        married        adethloff61@lycos.com   
474          nani lehmann  28.0       divorced              nlehmannd4@t.co   
1161            adey rome  55.0        married         arome4d@linkedin.com   
815         Tripp Yarnell  36.0      separated       tyarnellmk@oaic.gov.au   
1993       Stacee Eagland  61.0         single     seaglandrb@delicious.com   
1581        Ben Wrintmore  51.0         single       bwrintmorefy@naver.com   
1738   walliw stallybrass  28.0        married  wstallybrasska@geocities.jp   
1595    Hildegaard Glaves  57.0         single        hglavesgc@gizmodo.com   
948         Quintus Cavet  30.0       divorced             qcavetq9@mlb.com   
1638     ddene longbottom  51.0       divorced        dlongbottomhi@ucoz.ru   
1538      eadmund jotcham  26.0       divorced        ejotchamer@oakley.com   

             phone                                       full_address  \
861   609-240-7347                   8 Clove Alley,Trenton,New Jersey   
698   718-908-1120                     9 Becker Center,Bronx,New York   
260   501-872-1231      18 Bay Way,Hot Springs National Park,Arkansas   
1816  408-417-8990      5251 Stone Corner Parkway,San Jose,Kalifornia   
1371  910-707-2698  85531 Carpenter Crossing,Wilmington,NorthCarolina   
591   702-310-5574                   4 Logan Terrace,Las Vegas,Nevada   
835   202-135-7887  559 Fallview Circle,Washington,District of Col...   
1744  215-975-7963   295 Bunker Hill Street,Philadelphia,Pennsylvania   
1035  804-869-6072                530 Dapin Parkway,Richmond,Virginia   
217   202-797-1834  2 Stone Corner Terrace,Washington,District of ...   
474   915-860-6048                     967 Brown Circle,El Paso,Texas   
1161  806-769-4461                  0 Kensington Trail,Amarillo,Texas   
815   707-892-9585                  6 Dapin Court,Petaluma,California   
1993  513-794-5750                 246 Jenifer Center,Cincinnati,Ohio   
1581  361-400-7437         1 Meadow Ridge Circle,Corpus Christi,Texas   
1738  215-289-6324         66 Novick Street,Philadelphia,Pennsylvania   
1595  253-910-1247                 056 8th Junction,Tacoma,Washington   
948   202-381-0685  7591 Acker Avenue,Washington,District of Columbia   
1638  608-895-2738               454 Pearson Avenue,Madison,Wisconsin   
1538  850-870-5207                7265 Nancy Pass,Tallahassee,Florida   

                           job_title membership_date  
861             Safety Technician II        8/9/2015  
698                   Civil Engineer        9/2/2013  
260                 Media Manager II      12/26/2017  
1816               Assistant Manager       5/21/2017  
1371                Technical Writer       2/10/2019  
591      Information Systems Manager       11/4/2013  
835               Help Desk Operator      11/20/2020  
1744               Director of Sales       5/15/2019  
1035                    VP Marketing       12/1/2018  
217              Account Coordinator      12/25/2012  
474              Associate Professor       11/2/2014  
1161    Administrative Assistant III       6/10/2020  
815                Financial Analyst      11/22/2017  
1993    Structural Analysis Engineer       4/30/2017

#DataFrame Info

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2010 entries, 0 to 2009
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   full_name        2010 non-null   object 
 1   age              2007 non-null   float64
 2   martial_status   1990 non-null   object 
 3   email            2010 non-null   object 
 4   phone            2001 non-null   object 
 5   full_address     2010 non-null   object 
 6   job_title        1971 non-null   object 
 7   membership_date  2010 non-null   object 
dtypes: float64(1), object(7)
memory usage: 125.8+ KB


##Summary Statistik pada numerical columns

In [10]:
df.describe()

age
count  2007.000000
mean     44.926756
std      41.498191
min      18.000000
25%      32.000000
50%      41.000000
75%      51.000000
max     677.000000

- Rata Age pada rataan 44%
- Max Age 677 dugaan adanya outlier ataupun kesalahan data (cleansing)

#Missing Value Info

In [11]:
def find_missing_value(data_frame):
    missing_data = data_frame.isnull().sum(axis=0).reset_index()
    missing_data.columns = ['variable', 'missing values']
    missing_data['filling factor (%)'] = (data_frame.shape[0]-missing_data['missing values'])/data_frame.shape[0]*100
    return missing_data.sort_values('filling factor (%)').reset_index(drop = True)

find_missing_value(df)

variable  missing values  filling factor (%)
0        job_title              39           98.059701
1   martial_status              20           99.004975
2            phone               9           99.552239
3              age               3           99.850746
4        full_name               0          100.000000
5            email               0          100.000000
6     full_address               0          100.000000
7  membership_date               0          100.000000

Missing Value pada masing-masing Columns

#Cek Duplicates

In [12]:
duplicates = df[df.duplicated(keep=False)]
print(duplicates)

               full_name   age martial_status                       email  \
1323  Tamqrah Dunkersley  36.0         single   tdunkersley8u@dedecms.com   
1403  Tamqrah Dunkersley  36.0         single   tdunkersley8u@dedecms.com   
1480    Nicki Filliskirk  66.0        married  nfilliskirkd5@newsvine.com   
1600    Nicki Filliskirk  66.0        married  nfilliskirkd5@newsvine.com   
1840        ERWIN HUXTER  25.0        married   ehuxterm0@marketwatch.com   
1920        ERWIN HUXTER  25.0        married   ehuxterm0@marketwatch.com   

             phone                              full_address  \
1323  651-939-2423   0 Colorado Terrace,Saint Paul,Minnesota   
1403  651-939-2423   0 Colorado Terrace,Saint Paul,Minnesota   
1480  410-848-2272      7657 Alpine Plaza,Baltimore,Maryland   
1600  410-848-2272      7657 Alpine Plaza,Baltimore,Maryland   
1840  704-295-3261  0 Homewood Road,Charlotte,North Carolina   
1920  704-295-3261  0 Homewood Road,Charlotte,North Carolina   

           

Terdapat 3 duplicates dengan pengecekan dataframe menyeluruh

In [14]:
from pickle import FALSE
# Check for duplicates by specific columns
duplicate_details = df[df.duplicated(subset=['full_name'], keep=False)]

# Display the duplicate details
duplicate_details
duplicate_details.sort_values(by=['full_name'])


full_name    age martial_status                       email  \
1840        ERWIN HUXTER   25.0        married   ehuxterm0@marketwatch.com   
1800        ERWIN HUXTER   25.0         single   ehuxterm0@marketwatch.com   
1920        ERWIN HUXTER   25.0        married   ehuxterm0@marketwatch.com   
1254      Garrick Reglar   66.0         single       greglar4r@answers.com   
1175      Garrick Reglar   66.0      separated       greglar4r@answers.com   
1434      Haskell Braden  322.0       divorced      hbradenri@freewebs.com   
2000      Haskell Braden   32.0       divorced      hbradenri@freewebs.com   
1014    Maddie Morrallee   39.0       divorced  mmorralleemj@wordpress.com   
814     Maddie Morrallee   39.0        married  mmorralleemj@wordpress.com   
1480    Nicki Filliskirk   66.0        married  nfilliskirkd5@newsvine.com   
1600    Nicki Filliskirk   66.0        married  nfilliskirkd5@newsvine.com   
259      Obed MacCaughen   27.0        married     omaccaughen1o@naver.com   
60       Obed MacCaughen   27.0      separated     omaccaughen1o@naver.com   
450       Seymour Lamble   27.0         single      slamble81@amazon.co.uk   
290       Seymour Lamble   27.0        married      slamble81@amazon.co.uk   
1323  Tamqrah Dunkersley   36.0         single   tdunkersley8u@dedecms.com   
1403  Tamqrah Dunkersley   36.0         single   tdunkersley8u@dedecms.com   
803      georges prewett   47.0         single          gprewettfl@mac.com   
563      georges prewett   47.0        married          gprewettfl@mac.com   

             phone                                full_address  \
1840  704-295-3261    0 Homewood Road,Charlotte,North Carolina   
1800  704-295-3261    0 Homewood Road,Charlotte,North Carolina   
1920  704-295-3261    0 Homewood Road,Charlotte,North Carolina   
1254  214-314-5437                 6 Dottie Drive,Dallas,Texas   
1175  214-314-5437                 6 Dottie Drive,Dallas,Texas   
1434  510-963-9848  35005 Waubesa Crossing,Berkeley,California   
2000  510-963-9848  35005 Waubesa Crossing,Berkeley,California   
1014  712-853-2968        9339 Straubel Center,Sioux City,Iowa   
814   712-853-2968        9339 Straubel Center,Sioux City,Iowa   
1480  410-848-2272        7657 Alpine Plaza,Baltimore,Maryland   
1600  410-848-2272        7657 Alpine Plaza,Baltimore,Maryland   
259   815-990-8611      7069 Valley Edge Alley,Joliet,Illinois   
60    815-990-8611      7069 Valley Edge Alley,Joliet,Illinois   
450   979-346-7243               90691 Veith Place,Bryan,Texas   
290   979-346-7243               90691 Veith Place,Bryan,Texas   
1323  651-939-2423     0 Colorado Terrace,Saint Paul,Minnesota   
1403  651-939-2423     0 Colorado Terrace,Saint Paul,Minnesota   
803   571-157-7766         76 Graedel Road,Alexandria,Virginia   
563   571-157-7766         76 Graedel Road,Alexandria,Virginia   

                         job_title membership_date  
1840    Software Test Engineer III       9/29/2017  
1800    Software Test Engineer III       9/29/2017  
1920    Software Test Engineer III       9/29/2017  
1254           Safety Technician I      11/22/2015  
1175           Safety Technician I      11/22/2015  
1434              Dental Hygienist       11/4/2015  
2000              Dental Hygienist       11/4/2015  
1014          Software Engineer II       1/29/2020  
814           Software Engineer II       1/29/2020  
1480                  Geologist IV       6/18/2021  
1600                  Geologist IV       6/18/2021  
259               Junior Executive       1/27/2012  
60                Junior Executive       1/27/2012  
450   Budget/Accounting Analyst IV      12/26/2017  
290   Budget/Accounting Analyst IV      12/26/2017  
1323                      VP Sales       6/27/2016  
1403                      VP Sales       6/27/2016  
803                      Paralegal       10/5/2015  
563                      Paralegal       10/5/2015

Setelah pengecekan dupe berdasar nama terdapat 9 data Dupe

#Cek Numerical Columns

In [15]:
numeric_variable = df.select_dtypes(exclude=['object','category','bool']).columns
numeric_variable

Index(['age'], dtype='object')

In [16]:
numeric = df.loc[:, (df.dtypes == int) | (df.dtypes == float)].columns.tolist()
numeric

['age']

Numerical columns hanya [Age]

# Data Frame Copy

In [17]:
df_temp = df.copy()

#DF_Temp Cleaning

##Cleaning Name Column (extra space dan special char)

In [18]:
# def function Cleaning nama yang ada extra space dan special char dan di jadikan lowercase semua
def clean_full_name(name) :
  return  re.sub(r'[^a-zA-Z0-9\s]', '', name).strip().lower()

In [19]:
# apply on selected column
df_temp['full_name'] = df_temp['full_name'].apply(clean_full_name)
df_temp.sample(20)

full_name   age martial_status                         email  \
1612    bernete larvent  66.0        married          blarventgs@ameblo.jp   
545     billye heineken  53.0        married            bheinekenf3@va.gov   
561   lynelle ohartnett  55.0      separated  lohartnettfj@yellowpages.com   
1863    ramona macquist  59.0         single            rmacquistnq@w3.org   
27     garrik maestrini  48.0         single       gmaestrinir@addthis.com   
951         salomi feek  41.0         single         sfeekqc@wikipedia.org   
525         jobyna sirl  36.0         single                 jsirlej@ow.ly   
709      arney haldenby  25.0        married    ahaldenbyjn@technorati.com   
859          cart orteu  34.0         single      corteuns@theatlantic.com   
112         valera petz  41.0      separated             vpetz34@google.es   
1079      myrtle morman  45.0         single           mmorman23@phpbb.com   
626      gretta morrowe  45.0       divorced   gmorrowehc@businessweek.com   
1210     krispin mangon  32.0        married        kmangon5q@blogspot.com   
1910    vinnie shermore  36.0        married         vshermorep1@weibo.com   
17          kerr dorkin  41.0       divorced             kdorkinh@admin.ch   
1012      jecho bradden  56.0        married          jbradden9@nsw.gov.au   
121   rebecka langabeer  27.0        married       rlangabeer3d@apache.org   
1446    kendricks galey  36.0         single              kgaleyc7@wix.com   
589      peggie budgeon  35.0      separated         pbudgeongb@cpanel.net   
698     ardelis larmett  50.0         single      alarmettjc@blinklist.com   

             phone                                   full_address  \
1612  772-228-0535  9 Claremont Crossing,Port Saint Lucie,Florida   
545   602-459-1909             11503 Forster Lane,Phoenix,Arizona   
561   434-361-0198              31 2nd Parkway,Lynchburg,Virginia   
1863  217-167-2095            2470 1st Alley,Springfield,Illinois   
27    716-896-8482        8042 Continental Point,Buffalo,New York   
951   409-762-1358                 1 Butternut Court,Spring,Texas   
525   719-271-7031            49137 Memorial Lane,Denver,Colorado   
709   706-457-1431             02 Ridge Oak Court,Augusta,Georgia   
859   305-558-4547                50329 Hovde Way,Hialeah,Florida   
112   571-660-4990                18 Hoepker Pass,Vienna,Virginia   
1079  781-733-1210    6034 Hollow Ridge Road,Newton,Massachusetts   
626   702-279-3524      2572 Morningstar Parkway,Las Vegas,Nevada   
1210  717-665-7550       18 Sugar Parkway,Harrisburg,Pennsylvania   
1910  818-314-6026       22688 Dakota Place,Northridge,California   
17    702-560-2980              75 Basil Terrace,Las Vegas,Nevada   
1012  559-958-7175              71 Artisan Lane,Fresno,California   
121   914-542-6926         37567 Hauk Drive,Mount Vernon,New York   
1446  585-688-2794          8192 Michigan Hill,Rochester,New York   
589   608-843-7905        37176 Ridgeview Plaza,Madison,Wisconsin   
698   718-908-1120                 9 Becker Center,Bronx,New York   

                     job_title membership_date  
1612       Software Engineer I        7/1/2020  
545              Programmer II        8/5/2018  
561        Executive Secretary       8/21/2012  
1863         Marketing Manager        7/9/2018  
27           Financial Analyst       4/24/2018  
951                   VP Sales       6/12/2019  
525              Social Worker       1/10/2017  
709                     Editor       2/13/2021  
859                  Paralegal        4/8/2013  
112   Environmental Specialist       12/3/2019  
1079           Health Coach IV       2/10/2018  
626         Research Associate       1/11/2019  
1210                     Nurse       5/27/2019  
1910             Social Worker        4/3/2018  
17    Senior Financial Analyst       5/16/2021  
1012         Assistant Manager       6/15/2018  
121     Programmer Analyst III      10/28/2017  
1446       Office Assistant II      10/26/2

##Cleaning Age Column ('Na' , Change Type to Int, Truncate Age digit)

In [20]:
#Ubah [age] from float to int
df_temp['age'] = df_temp['age'].astype('Int64')

In [21]:
#Karena missing value [age] hanya 3 maka di drop saja karena tidak akan mempengaruhi data
df_temp = df_temp.dropna(subset=['age'])

In [ ]:
#Truncate 3 Digit age jadi 2 di awal
df_temp['age'] = df_temp['age'].apply(lambda x: int(str(int(x))[:2]) if x >= 100 else int(x))

## Cleaning Married Status Column (typo status, Fill NA)

In [24]:
df_temp['martial_status'] = df_temp['martial_status'].str.replace('divored', 'divorced')

In [25]:
cleaner = SimpleImputer(strategy='constant', fill_value = 'Null')
df_temp['martial_status'] = cleaner.fit_transform(df_temp['martial_status'].values.reshape(-1, 1)).flatten()

## Cleaning Phone Number Column (Fill Na, Replace invalid number)

In [26]:
cleaner = SimpleImputer(strategy='constant', fill_value = 'Nan')
df_temp['phone'] = cleaner.fit_transform(df_temp['phone'].values.reshape(-1, 1)).flatten()

In [27]:
#Phone Number <12
def validate_phone(phone):
    if len(str(phone)) < 12:
        return np.nan
    return phone

In [28]:
df_temp['phone'] = df_temp['phone'].apply(validate_phone)

##Cleaning Job Tittle (Fill Na, Trim Roman Number)

In [29]:
cleaner = SimpleImputer(strategy='constant', fill_value = 'Null')
df_temp['job_title'] = cleaner.fit_transform(df_temp['job_title'].values.reshape(-1, 1)).flatten()

In [30]:
# def function Cleaning nama yang ada extra space dan special char dan di jadikan lowercase semua
def job_title_clean(job) :
  return  re.sub(r'\b[IVXLCDM]+\b', '', job).strip()

In [31]:
df_temp['job_title'] = df_temp['job_title'].apply(job_title_clean)
df_temp.sample(20)

full_name  age martial_status                        email  \
615       gordon carlyon   40         single      gcarlyonh1@e-recht24.de   
1302       hilton bessom   25      separated    hbessom89@livejournal.com   
1808       terrijo wyett   21         single           twyettm8@desdev.cn   
637      averill maywood   48        married            amaywoodhn@cbc.ca   
19       geoffry bouette   33        married           gbouettej@live.com   
1336    humfrid noddings   65         single    hnoddings97@sitemeter.com   
1606    jermaine kindell   41        married         jkindellgm@intel.com   
1881          egor stihl   45         single     estihlo8@dailymail.co.uk   
1383    leandra miquelet   46        married         lmiqueletai@time.com   
874     babbette jeacock   25       divorced     bjeacocko7@xinhuanet.com   
1062      huey van merwe   20         single               hvan1m@free.fr   
1297       darcy jozwiak   67        married       djozwiak84@vinaora.com   
1526    hubie chadderton   24        married      hchaddertonef@goo.ne.jp   
1415       rustin hotton   38         single              rhottonbd@51.la   
2007     phillip piddick   40        married          ppiddickrp@hibu.com   
758         cassie greep   27        married         cgreepl0@sina.com.cn   
785     gonzalo archbold   45        married  garchboldlr@arstechnica.com   
1838        martie crumb   47        married          mcrumbn2@taobao.com   
113   milissent kniveton   53        married      mkniveton35@archive.org   
1584      karyl mallabon   27        married       kmallabong1@elpais.com   

             phone                                 full_address  \
615   865-863-0874       0505 Hansons Drive,Knoxville,Tennessee   
1302  901-567-9450      9515 Washington Plaza,Memphis,Tennessee   
1808  301-483-1582    86 High Crossing Trail,Baltimore,Maryland   
637   334-511-4945   9 Crescent Oaks Parkway,Montgomery,Alabama   
19    361-160-6496          53 Knutson Way,Corpus Christi,Texas   
1336  210-707-8495              003 Almo Lane,San Antonio,Texas   
1606  828-402-6291  621 Fallview Drive,Asheville,North Carolina   
1881  281-893-8298                25 Pankratz Pass,Humble,Texas   
1383  703-655-5817       1386 Bobwhite Trail,Arlington,Virginia   
874   925-140-6288          9 Sherman Avenue,Concord,California   
1062  850-161-1680      05525 Arrowood Avenue,Pensacola,Florida   
1297  601-747-9163        8 Gulseth Terrace,Jackson,Mississippi   
1526  508-812-4769       156 Scott Point,Brockton,Massachusetts   
1415  602-179-0278       06 Crownhardt Drive,Scottsdale,Arizona   
2007  512-886-9162        767 Burning Wood Parkway,Austin,Texas   
758   517-153-3171          24017 Hanson Point,Lansing,Michigan   
785   212-981-8341       23 Cody Terrace,New York City,New York   
1838  312-923-4744           5940 Kennedy Pass,Chicago,Illinois   
113   469-861-5680              792 Oakridge Court,Dallas,Texas   
1584  901-103-0932            9291 Lawn Place,Memphis,Tennessee   

                          job_title membership_date  
615                Technical Writer       10/1/2020  
1302              Software Engineer       12/5/2017  
1808            Executive Secretary       7/28/2021  
637                Product Engineer      10/14/2012  
19            Systems Administrator      11/19/2014  
1336         Occupational Therapist        1/9/2020  
1606            Associate Professor      11/15/2018  
1881       Computer Systems Analyst        7/1/2021  
1383  Analog Circuit Design manager      11/27/2015  
874        Environmental Specialist      11/28/2019  
1062           Accounting Assistant      11/23/2021  
1297        Human Resources Manager      10/23/2018  
1526         Senior Sales Associate       9/25/2016  
1415                   Food Chemist        4/4/2021  
2007             Research Assistant      10/18/2016  
758                Dental Hygienist      10/26/2014  
785                Product Engineer       5/21/2021  
1838          Aut

##Cleaning Membership Date Column (replace year < 2000)

In [32]:
df_temp['membership_date'] = pd.to_datetime(df_temp['membership_date'])

In [33]:
df_temp['membership_date'] = df_temp['membership_date'].apply(lambda x: x.replace(year=2000) if x.year < 2000 else x)

##Missing Value DF_Temp

In [34]:
def find_missing_value(data_frame):
    missing_data = data_frame.isnull().sum(axis=0).reset_index()
    missing_data.columns = ['variable', 'missing values']
    missing_data['filling factor (%)'] = (data_frame.shape[0]-missing_data['missing values'])/data_frame.shape[0]*100
    return missing_data.sort_values('filling factor (%)').reset_index(drop = True)

find_missing_value(df_temp)

variable  missing values  filling factor (%)
0            phone              14           99.302441
1        full_name               0          100.000000
2              age               0          100.000000
3   martial_status               0          100.000000
4            email               0          100.000000
5     full_address               0          100.000000
6        job_title               0          100.000000
7  membership_date               0          100.000000

##Remove Duplicate

In [35]:
df_temp = df_temp.drop_duplicates(subset=['full_name'],keep = 'first')

remove dupe sort by dupe name, keep first

In [36]:
df_temp.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1997 entries, 0 to 2009
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   full_name        1997 non-null   object        
 1   age              1997 non-null   int64         
 2   martial_status   1997 non-null   object        
 3   email            1997 non-null   object        
 4   phone            1983 non-null   object        
 5   full_address     1997 non-null   object        
 6   job_title        1997 non-null   object        
 7   membership_date  1997 non-null   datetime64[ns]
dtypes: datetime64[ns](1), int64(1), object(6)
memory usage: 140.4+ KB
